### PR TITLE
WIP Native Events

### LINKED ISSUE #311: Support for Native Browser Events
### click

- [x] simulated click fixes #2956 
- [x] on text-editable click in center move the cursor to the end
  - [x] ensure possible in `IE11` without causing side effects
- [x] `cy.body().type()`:
  - `blur` focused element, send events to `body`
- [x] support `.rightClick()` :new: 

#### possible:
- [ ] move cursor to beginning if `click('left')` or `click('topleft')` :warning: 
#### not fixed:
- [ ] document.execCommand("copy") does not work in cypress #2851 (execCommand will not work with untrusted events)

### type
- [x] by default issues native events
- [x] `force:true` skips actionability and uses simulated keys
- [ ] simulated keys have no delay :warning: change from 50ms
- [x] native keys by default have no delay :warning: change from 50ms
- [x] support new special `.type()` sequences: `{moveToEnd}`,`{moveToStart}` :new:
- [x] support new special `.type()` sequences: `{tab}`,`{shift+tab}` :new:  
- [x] type follows focus :warning: 
- [x] trusted native keyboard events :new: 
  - checkValidity() returns true instead of false for min length on input #1930
<details markdown><summary><code>
{moveToEnd}/{moveToStart}</code></summary>

- IE: setSelectionRange on all inputs
- chrome/ff: setSelectionRange except for email/number inputs. For those we use `execCommand('selectAll') + selection.modify('...')`
- branch logic based on `browserFeatures` object. For ex)
```js
{
  setSelectionRangeOnAllInputTypes: true
}
```

</details>

### focus
- `.focus` behavior on certain text editable elements
  - [ ] command should first focusable element :warning: change from subject
  - [ ] focus host contenteditable #2717
### contenteditable
- treated as any other text-editable in the commands above. 
- By default move cursor to end
- [x] support `{moveToEnd}`/`{moveToStart}`
- [x] support `{tab}`

### file uploads :new: 
- proposal https://github.com/cypress-io/cypress/issues/170
- [ ] tbd

### cy.hover :new:
- proposal https://github.com/cypress-io/cypress/issues/10
- [ ] tbd

### scroll behavior
- [ ] scroll into view only if needed :warning: 
  - use getElementFromPoint on the pixel needing click
- [ ] scroll into center by default :warning: 
- [ ] allow user to configure scroll position (center, bottom, top) Enable users to change the scrolling strategy #871 :new: 
**For IE:**
  - scrollIntoView can't center, so calculate coords and try to center it manually

### select text command? :new: 
- allow user to make arbitrary text selection

### mouse state
- [ ] tbd

### ~for mouse actions that involve mouse state:~
- in open mode, warn if potentially affected test
- includes click (since hover before click), hover, drag-and-drop

For IE11 & firefox, see branch [issue-311-ie-ff](https://github.com/cypress-io/cypress/tree/issue-311-ie-ff)

### LINKED ISSUE #2613: Cannot type on datetime-local input
Hi, i'm trying to type on an input of `datetime-local` type.
On the documentation, it says that `datetime-local` is supported but nothing was never typed, no matter how I tried to type it `type.('10102019{rightarrow1010pm}'`,`type.('000000000000pm}')`. 
Also, in this #1366 issue has a workaround. but both did not work. 
It does select the input, but never types.

### LINKED ISSUE #3405: type('{enter}') doesn't fire input event
### Current behavior:

I'm trying to test an app with the ace editor ( https://ace.c9.io/ ), which uses a `textarea`, but I'm running into a problem:

```js
describe('textarea', () => {
  it('can type the enter key', () => {
    cy.visit('http://ace.c9.io/build/kitchen-sink.html');
    cy.get('#editor-container .ace_editor textarea').type('foo{enter}', {force: true});
  });
});

```

It types 'foo' but does not press the Enter key.

### Desired behavior:

Using `{enter}` in `type` should press the Enter key.

### Steps to reproduce: (app code and test code)

See above.

### Versions

3.1.5, Chrome 72, MacOS

### LINKED ISSUE #3722: cy.type doesn't fill in KeyboardEvent.code
### Current behavior:

The `KeyboardEvent` object generated when using the `type` command doesn't contain the `code` property.

### Desired behavior:

The `KeyboardEvent` object generated when using the `type` command should contain the `code` property.

https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code

### Versions

3.1.5

### LINKED ISSUE #4767: cannot `type` negative number into number input with text selection
Cannot `type` negative number into number input with text selection

Relevant issue description: https://github.com/cypress-io/cypress/issues/2173#issuecomment-513099254

### PR DESCRIPTION
**Pull request by kuceb**
fix #311 

- typing into datetime-local input type
fix #2613

- input events for `{enter}`: 
fix #3405
- sending `code` property on keyboard event
fix #3722
- [ ] follow w3 spec for number inputs
fix #4767
### Tasks
- [x] waiting on #4103 use webpack internally
- [x] waiting on #2956 simulated click fixes
- [ ] pull in changes from develop + type_spec.coffee
- [ ] factor out synthetic type changes

### CODE REVIEW

File: packages/driver/src/cy/actionability.coffee [hunk 1]
```diff
@@ -186,6 +186,7 @@ ensureElIsNotCovered = (cy, win, $el, fromViewport, options, log, onScroll) ->
   return $elAtCoords
 
 getCoordinatesForEl = (cy, $el, options) ->
+  debugger
```
Comment (jennifer-shehane) [inline packages/driver/src/cy/actionability.coffee]: Y'all need some linters running on precommit hooks, debugger should never been committed. 👀

Comment (Bkucera) [reply packages/driver/src/cy/actionability.coffee]: its a WIP, so i use `--no-verify`

Comment (Bkucera) [reply packages/driver/src/cy/actionability.coffee]: I don't think we should bother reviewing WIP PR's, isn't that the point of WIP? :smile:

Comment (brian-mann) [reply packages/driver/src/cy/actionability.coffee]: Agree with @Bkucera 	if we indicate a WIP on a PR it should be ignored and not be reviewed.

**Comment by CypressJoseph** [comment]
@Bkucera any objections to closing this one?

**Comment by Bkucera** [comment]
closing since this is very stale and no longer a blocker for cross-browser support (which is released)

**Pull request closed without merging**